### ISSUE TITLE
Camel case to snake case name conversion

### ISSUE THREAD

**Issue by joshiain** [supabase/postgrest-js#204]
# Feature request

## Is your feature request related to a problem? Please describe.

postgres column naming conventions suggest using snake_case. Javascript variable naming conventions typically use camelCase. This leads to a few issues when trying to interact with the postgres database using the postgrest-js api.

I either have to use snake_case for my object keys in my javascript code, but this is then inconsistent with the rest of my code. 

I could also use camelCase for my postgres column names, but then this creates a lot of friction with queries as casing is ignored without "".

Or I could automatically convert the keys from camelCase to snake_case before sending any data, and from snake_case to camelCase when receiving data.

## Describe the solution you'd like

A config setting, or option that will automatically convert object keys from camelCase to snake_case when sending data.

Having a table with columns: `user_name` and `date_added`. The following would work:
```javascript
const data = [{userName: "abc123", dateAdded: "2021-01-01"}]
await postgrest.select('table').insert(data)
```

And similarly when querying for data, the returned object would have the column names converted from snake_case to camelCase.

## Describe alternatives you've considered

pg-promise, knex.js, and objection.js provide similar utils to help with this issue.

**Comment by wiesson** [comment]
This would be soooo helpful :)

**Comment by steve-chavez** [comment]
[Renaming columns](https://postgrest.org/en/stable/api.html#renaming-columns) on reads is possible, but not on writes(https://github.com/PostgREST/postgrest/issues/1773) - once the latter is solved `postgrest-js` could autogenerate aliases to solve this.

**Comment by wiesson** [comment]
I ended up with converting my db to the camel case format and accepted that I have to live with quotes forever

**Comment by steve-chavez** [comment]
Another option for this: https://github.com/supabase/supabase/discussions/7136

**Comment by irohitb** [comment]
If this helps anyone, I have been using [`ts-case-convert`](https://www.npmjs.com/package/ts-case-convert) and it works like magic.

**Comment by weeksie** [comment]
This is extremely frustrating. It's a very simple move to add a conversion. Knex does it without sweating. The "accepted" option is basically to break encapsulation so that you can tell you have a data object b/c its properties are in snake case. I appreciate "opinionated" frameworks as much as the next guy but a typescript/javascript framework (of all things) should have this affordance.

**Comment by wiesson** [comment]
Offtopic: I'm also hoping that at some point, it will be added to postgrest. It feels like a feature were I would apply for a job, fix that one thing and leave 🙈

**Comment by alvgaona** [comment]
+1

**Comment by isaachinman** [comment]
Can anyone here elaborate as to how you are using `ts-case-convert` to automatically wrap methods?

**Comment by joryphillips** [comment]
@isaachinman you probably figured this out a while ago, but this is more or less it. You'd want to do the reverse for write operations. 

```
import { objectToCamel } from "ts-case-convert";

export async function getThingById(id: string) {
  const tableQuery = supabase
    .from("my_cool_table")
    .select("*")
    .eq("id", id);

  const { data, error } = await tableQuery;
  if (error) throw error;

  return objectToCamel(data);
}
```

**Comment by veloware** [comment]
> If this helps anyone, I have been using [`ts-case-convert`](https://www.npmjs.com/package/ts-case-convert) and it works like magic.

Ive gave this a go, and it works, but I'm curious if you ran into any complexities regarding inferring the TS types? as in - 

normally you could do - 

```
const myQuery = supabase.from("xxx")...

type MyQueryResponse = QueryData<typeof myQuery>;
```

and then `MyQueryResponse` would have all of the **snake_case** properties, but im curious if you know how to infer types after the `objectToCamel` conversion has happened?

**Comment by Revadike** [comment]
> [Renaming columns](https://postgrest.org/en/stable/api.html#renaming-columns) on reads is possible, but not on writes([PostgREST/postgrest#1773](https://github.com/PostgREST/postgrest/issues/1773)) - once the latter is solved `postgrest-js` could autogenerate aliases to solve this.

I don't think we need this. We can do the conversion before sending the postgREST calls.

**Comment by Revadike** [comment]
^ For this reason I recommend @steve-chavez removing the "postgrest-patch-needed" label

**Comment by steve-chavez** [comment]
@Revadike So you mean we should delegate this to `ts-case-convert`? 🤔 

Would there be any disadvantages compared to server side support? Like, could you get mysterious error messages?

**Comment by alvgaona** [comment]
`ts-case-convert` is definitely not the way to go to provide a great DX.

**Comment by Revadike** [comment]
> [@Revadike](https://github.com/Revadike) So you mean we should delegate this to `ts-case-convert`? 🤔
> 
> Would there be any disadvantages compared to server side support? Like, could you get mysterious error messages?

No, if this option is enabled by the user, literally just convert all the keys they provide into snake case, and all the keys the that are retrieved into camelcase.
